linker/shared/src/main/scala/org/scalajs/linker/analyzer/Analyzer.scala

```diff
@@ -1572,22 +1572,29 @@ private object AnalyzerRun {
 
   private class WorkTracker(implicit ec: ExecutionContext) {
     private val pending = new AtomicInteger(0)
+    private val failures = new AtomicReference[List[Throwable]](Nil)
     @volatile private var _allowComplete = false
     private val promise = Promise[Unit]()
 
     def track(fut: Future[Unit]): Unit = {
       pending.incrementAndGet()
 
-      fut.onComplete {
-        case Success(_) =>
-          if (pending.decrementAndGet() == 0)
-            tryComplete()
-
-        case Failure(t) =>
-          promise.tryFailure(t)
+      fut.onComplete { result =>
+        result match {
+          case Success(_) => ()
+          case Failure(t) => addFailure(t)
+        }
+        if (pending.decrementAndGet() == 0)
+          tryComplete()
       }
     }
 
+    @tailrec
+    private def addFailure(t: Throwable): Boolean = {
+      val prev = failures.get()
+      failures.compareAndSet(prev, t :: prev) || addFailure(t)
```

Comment:
I find the use of `||` a bit funky here, especially since it leaks into the return type.

```suggestion
      if (!failures.compareAndSet(prev, t :: prev))
        addFailure(t)
```

Reply:
Done.